### PR TITLE
Added `databricks labs ucx validate-external-locations` command for cli

### PR DESCRIPTION
**Pull request by HariGS-DB**
Description:
This change contains steps to validate external location defined in the external tables with the ones already available in the unity external location. It prints the count of tables that can be migrated for external location already present. Also shares the list of external location that needs to be created for the missing ones. It generates a tf file

Changes:
labs.yml - to add the new command
cli.py - to add code for invoking the new command
locations.py - add a field which gives count of tables using the external location. also added logic to identify duplicate for jdbc connection objects, functionality to compare locations and print details and generate tf file
updated test cases to locations
added new test cases for mapping

### CODE REVIEW

File: src/databricks/labs/ucx/hive_metastore/mapping.py [hunk 1]
```diff
@@ -106,3 +107,62 @@ def skip_schema(self, backend: StatementExecutionBackend, schema: str):
...
+            folder = f"/Users/{ws.current_user.me().user_name}/.ucx"
+        self._ws = ws
+        self._folder = folder
+
+    def _get_ext_location_definitions(self, missing_locations: list[ExternalLocation]) -> list:
+        tf_script = []
+        cnt = 1
+        for loc in missing_locations:
+            script = f'resource "databricks_external_location" "name_{cnt}" {{ \n'
+            script += f'name = "name_{cnt}"\n'
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/mapping.py:124]: insert four spaces before field names

File: src/databricks/labs/ucx/hive_metastore/mapping.py [hunk 2]
```diff
@@ -106,3 +107,62 @@ def skip_schema(self, backend: StatementExecutionBackend, schema: str):
...
+        if not folder:
+            folder = f"/Users/{ws.current_user.me().user_name}/.ucx"
+        self._ws = ws
+        self._folder = folder
+
+    def _get_ext_location_definitions(self, missing_locations: list[ExternalLocation]) -> list:
+        tf_script = []
+        cnt = 1
+        for loc in missing_locations:
+            script = f'resource "databricks_external_location" "name_{cnt}" {{ \n'
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/mapping.py:123]: use semantic names, you can derive them from bucket names or account names

File: src/databricks/labs/ucx/hive_metastore/mapping.py [hunk 3]
```diff
@@ -106,3 +107,62 @@ def skip_schema(self, backend: StatementExecutionBackend, schema: str):
...
+            script += f'name = "name_{cnt}"\n'
+            script += f'url  = "{loc.location}"\n'
+            script += "credential_name = <storage_credential_reference>\n"
+            script += "}\n"
+            tf_script.append(script)
+            cnt += 1
+        return tf_script
+
+    def _match_table_external_locations(self, locations: ExternalLocations) -> tuple[list, list[ExternalLocation]]:
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/mapping.py:132]: i think it's better to put this logic into `ExternalLocations` class.

File: src/databricks/labs/ucx/hive_metastore/mapping.py [hunk 4]
```diff
@@ -106,3 +107,62 @@ def skip_schema(self, backend: StatementExecutionBackend, schema: str):
...
+        matching_locations = []
+        missing_locations = []
+        for loc in table_locations:
+            if loc.location in location_path:
+                matching_locations.append([external_locations[location_path.index(loc.location)].name, loc.table_count])
+                continue
+            missing_locations.append(loc)
+        return matching_locations, missing_locations
+
+    def save(self, locations: ExternalLocations) -> str:
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/mapping.py:145]: this method would also be better on `ExternalLocations`.

File: tests/unit/hive_metastore/test_mapping.py [hunk 5]
```diff
@@ -131,3 +140,31 @@ def test_skip_missing_table(mocker, caplog):
...
+    ws = MagicMock()
+    ext_location_mapping = ExternalLocationMapping(ws, "~/.ucx")
+    location_crawler = create_autospec(ExternalLocations)
+    location_crawler.snapshot.return_value = [ExternalLocation(location="s3://test_location/test1", table_count=1)]
+    ws.external_locations.list.return_value = [ExternalLocationInfo(name="loc1", url="s3://test_location/test11")]
+    ext_location_mapping.save(location_crawler)
+    (path, content), _ = ws.workspace.upload.call_args
+    assert "~/.ucx/external_locations.tf" == path
+    assert (
+        'resource "databricks_external_location" "name_1" { \n'
```
Comment (nfx) [inline tests/unit/hive_metastore/test_mapping.py:155]: e.g. the auto-created resource name here should be `test_location`

File: src/databricks/labs/ucx/hive_metastore/mapping.py [hunk 6]
```diff
@@ -106,3 +107,62 @@ def skip_schema(self, backend: StatementExecutionBackend, schema: str):
...
+        self._folder = folder
+
+    def _get_ext_location_definitions(self, missing_locations: list[ExternalLocation]) -> list:
+        tf_script = []
+        cnt = 1
+        for loc in missing_locations:
+            script = f'resource "databricks_external_location" "name_{cnt}" {{ \n'
+            script += f'name = "name_{cnt}"\n'
+            script += f'url  = "{loc.location}"\n'
+            script += "credential_name = <storage_credential_reference>\n"
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/mapping.py:126]:
```suggestion
            script += "credential_name = databricks_storage_credential.<storage_credential_reference>.id\n"
```

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/cli.py [hunk 7]
```diff
@@ -90,13 +90,31 @@ def create_table_mapping():
...
+    ws = WorkspaceClient()
+    prompts = Prompts()
+    installation_manager = InstallationManager(ws)
+    installation = installation_manager.for_user(ws.current_user.me())
+    sql_backend = StatementExecutionBackend(ws, installation.config.warehouse_id)
+    location_crawler = ExternalLocations(ws, sql_backend, installation.config.inventory_database)
+    path = location_crawler.save()
+    if len(path) > 0:
+        return
+    if (
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:103]: `if prompts.confirm(...`

File: src/databricks/labs/ucx/hive_metastore/locations.py [hunk 8]
```diff
@@ -113,6 +127,71 @@ def _try_fetch(self) -> Iterable[ExternalLocation]:
...
+        for loc in table_locations:
+            if loc.location[:-1].lower() in location_path:
+                matching_locations.append(
+                    [external_locations[location_path.index(loc.location[:-1])].name, loc.table_count]
+                )
+                continue
+            missing_locations.append(loc)
+        return matching_locations, missing_locations
+
+    def save(self) -> str:
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/locations.py:170]:
```suggestion
    def save_as_terraform_definitions_on_workspace(self) -> str:
```
remove folder argument from constructor, it doesn't belong there, pass it as the argument for this method.

File: src/databricks/labs/ucx/hive_metastore/locations.py [hunk 9]
```diff
@@ -113,6 +127,71 @@ def _try_fetch(self) -> Iterable[ExternalLocation]:
...
+            script = f'resource "databricks_external_location" "{res_name}" {{ \n'
+            script += f'    name = "{res_name}"\n'
+            script += f'    url  = "{loc.location[:-1]}"\n'
+            script += "    credential_name = databricks_storage_credential.<storage_credential_reference>.id\n"
+            script += "}\n"
+            tf_script.append(script)
+            cnt += 1
+        return tf_script
+
+    def _match_table_external_locations(self) -> tuple[list, list[ExternalLocation]]:
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/locations.py:155]: what's the type of the first return argument in a tuple?..

File: src/databricks/labs/ucx/hive_metastore/locations.py [hunk 10]
```diff
@@ -113,6 +127,71 @@ def _try_fetch(self) -> Iterable[ExternalLocation]:
...
+            elif loc.location.startswith("gcs://"):
+                res_name = loc.location[6:].replace("/", "_")[:-1]
+            elif loc.location.startswith("abfss://"):
+                res_name = (
+                    loc.location[loc.location.index("@") + 1 :]
+                    .replace(".dfs.core.windows.net", "")
+                    .replace("/", "_")[:-1]
+                )
+            else:
+                res_name = f"name_{cnt}"
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/locations.py:145]: it's unclear where this happens. please add a comment and log something

File: src/databricks/labs/ucx/hive_metastore/locations.py [hunk 11]
```diff
@@ -113,6 +127,71 @@ def _try_fetch(self) -> Iterable[ExternalLocation]:
...
+        return tf_script
+
+    def _match_table_external_locations(self) -> tuple[list, list[ExternalLocation]]:
+        external_locations = list(self._ws.external_locations.list())
+        location_path = [_.url.lower() for _ in external_locations]
+        table_locations = self.snapshot()
+        matching_locations = []
+        missing_locations = []
+        for loc in table_locations:
+            if loc.location[:-1].lower() in location_path:
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/locations.py:162]: what does `loc.location[:-1]` do?.. I have difficulty reading this code, introduce more variables.

File: src/databricks/labs/ucx/hive_metastore/locations.py [hunk 12]
```diff
@@ -113,6 +127,71 @@ def _try_fetch(self) -> Iterable[ExternalLocation]:
...
+                matching_locations.append(
+                    [external_locations[location_path.index(loc.location[:-1])].name, loc.table_count]
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/locations.py:164]: I have difficulty reading this code, introduce more variables.

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/hive_metastore/locations.py (hunk 9 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/hive_metastore/locations.py:155]: its a combination of str and count. name from ws.external location and table count from ExternalLocation

File: src/databricks/labs/ucx/hive_metastore/locations.py [hunk 13]
```diff
@@ -30,6 +33,8 @@ class ExternalLocations(CrawlerBase[ExternalLocation]):
     def __init__(self, ws: WorkspaceClient, sbe: SqlBackend, schema):
         super().__init__(sbe, "hive_metastore", schema, "external_locations", ExternalLocation)
         self._ws = ws
+        folder = f"/Users/{ws.current_user.me().user_name}/.ucx"
+        self._folder = folder
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/locations.py:37]: Nit: inline a variable, it's used only once to assign a property

File: src/databricks/labs/ucx/hive_metastore/locations.py [hunk 14]
```diff
@@ -113,6 +126,78 @@ def _try_fetch(self) -> Iterable[ExternalLocation]:
...
+                res_name = loc.location[6:].replace("/", "_")[:-1]
+            elif loc.location.startswith("abfss://"):
+                res_name = (
+                    loc.location[loc.location.index("@") + 1 :]
+                    .replace(".dfs.core.windows.net", "")
+                    .replace("/", "_")[:-1]
+                )
+            else:
+                # if the cloud storage url doesn't match the above condition or incorrect (example wasb://)
+                # generate a dummy resource name
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/locations.py:145]: Wasb is not supported by UC ext locations. Raise this as warning and don't include this location in the list of tf resources

File: src/databricks/labs/ucx/hive_metastore/locations.py (hunk 9 above)
Comment (nfx) [reply src/databricks/labs/ucx/hive_metastore/locations.py:157]: Why isn't this clear from a type?

File: src/databricks/labs/ucx/hive_metastore/locations.py [hunk 15]
```diff
@@ -113,6 +126,78 @@ def _try_fetch(self) -> Iterable[ExternalLocation]:
         for row in self._fetch(f"SELECT * FROM {self._schema}.{self._table}"):
             yield ExternalLocation(*row)
 
+    def _get_ext_location_definitions(self, missing_locations: list[ExternalLocation]) -> list:
+        tf_script = []
+        cnt = 1
+        for loc in missing_locations:
+            if loc.location.startswith("s3://"):
+                res_name = loc.location[5:].replace("/", "_")[:-1]
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/locations.py:134]: Can this (and following) lines be rewritten with .rstrip()?

File: src/databricks/labs/ucx/hive_metastore/locations.py [hunk 16]
```diff
@@ -113,6 +126,78 @@ def _try_fetch(self) -> Iterable[ExternalLocation]:
...
+            for _ in missing_locations:
+                logger.info(f"{_.table_count} tables can be migrated using external location {_.location}.")
+            buffer = io.StringIO()
+            for script in self._get_ext_location_definitions(missing_locations):
+                buffer.write(script)
+            buffer.seek(0)
+            return self._overwrite_mapping(buffer)
+        else:
+            logger.info("no additional external location to be created.")
+            return str(None)
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/locations.py:194]: This is a bug. The calling code expects a string, not None.

File: tests/unit/hive_metastore/test_locations.py [hunk 17]
```diff
@@ -99,3 +109,61 @@ def test_external_locations():
...
+        "    credential_name = databricks_storage_credential.<storage_credential_reference>.id\n"
+        "}\n"
+        'resource "databricks_external_location" "test_location2_test2" { \n'
+        '    name = "test_location2_test2"\n'
+        '    url  = "gcs://test_location2/test2"\n'
+        "    credential_name = databricks_storage_credential.<storage_credential_reference>.id\n"
+        "}\n"
+        'resource "databricks_external_location" "storagetest1_test2" { \n'
+        '    name = "storagetest1_test2"\n'
+        '    url  = "abfss://cont1@storagetest1.dfs.core.windows.net/test2"\n'
```
Comment (nfx) [inline tests/unit/hive_metastore/test_locations.py:150]: Bug: container name is not part of automated name inference. This breaks when many containers are in the same storage account

**Review by nfx — CHANGES_REQUESTED** [review]
Fix bugs and change the PR title to mention the full command as users would invoke it.

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**